src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Utilities
+{
+    /// <summary>
+    /// Extensions to <see cref="CancellationToken"/>.
+    /// </summary>
+    internal static class CancellationTokenExtensions
+    {
+        /// <summary>
+        /// Creates a new <see cref="CancellationToken"/> that is canceled when any of a set of other tokens are canceled.
+        /// </summary>
+        /// <param name="original">The first token.</param>
+        /// <param name="other">The second token.</param>
+        /// <returns>A struct that contains the combined <see cref="CancellationToken"/> and a means to release memory when you're done using it.</returns>
+        public static CombinedCancellationToken CombineWith(this CancellationToken original, CancellationToken other)
+        {
+            if (original.IsCancellationRequested || !other.CanBeCanceled)
+            {
+                return new CombinedCancellationToken(original);
+            }
+
+            if (other.IsCancellationRequested || !original.CanBeCanceled)
+            {
+                return new CombinedCancellationToken(other);
+            }
```

Comment:
dumb question.  why do you need to check the .CanBeCanceled part?  if you already know that one or the other has had cancellation requested, why do you ever need to check the other.

The docs say "Creates a new <see cref="CancellationToken"/> that is canceled when **any** of a set of other tokens are canceled."  Since you already know "**any**" token was canceled, why are the others relevant at all?

Reply:
If `original` cannot be cancelled, then `CancellationTokenSource.CreateLinkedSource(other, original).Token` is observably equivalent to `other.Token`, so we avoid the allocation. The same is true when `other` is already cancelled. Based on the wording of your question, it sounds like you may have misread the condition as `&&` when it's `||`. Can you see if this answers your questions?

Reply:
Indeed!  I did misread it in that exact manner.  Thanks for the clarification!

Reply:
Retracted.  Misread things